numpy/core/src/multiarray/dtypemeta.c

```diff
@@ -0,0 +1,269 @@
+/* Array Descr Object */
+
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+#include "structmember.h"
+#include "assert.h"
+
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+#define _MULTIARRAYMODULE
+#include <numpy/ndarraytypes.h>
+#include "npy_pycompat.h"
+
+#include "dtypemeta.h"
+
+
+static void
+dtypemeta_dealloc(PyArray_DTypeMeta *self) {
+    /* Do not accidentally delete a statically defined DType: */
+    assert(((PyTypeObject *)self)->tp_flags & Py_TPFLAGS_HEAPTYPE);
+
+    Py_XDECREF(self->scalar_type);
+    Py_XDECREF(self->singleton);
+    PyType_Type.tp_dealloc((PyObject *) self);
+}
+
+static PyObject *
+dtypemeta_new(PyTypeObject *NPY_UNUSED(type),
+        PyObject *NPY_UNUSED(args), PyObject *NPY_UNUSED(kwds))
+{
+    PyErr_SetString(PyExc_TypeError,
+            "Preliminary-API: Cannot subclass DType.");
+    return NULL;
+}
+
+static int
+dtypemeta_init(PyTypeObject *NPY_UNUSED(type),
+        PyObject *NPY_UNUSED(args), PyObject *NPY_UNUSED(kwds))
+{
+    PyErr_SetString(PyExc_TypeError,
+            "Preliminary-API: Cannot __init__ DType class.");
+    return -1;
+}
+
+/**
+ * tp_is_gc slot of Python types. This is implemented only for documentation
+ * purposes to indicate and document the subtleties involved.
+ *
+ * Python Type objects are either statically created (typical C-Extension type)
+ * or HeapTypes (typically created in Python).
+ * HeapTypes have the Py_TPFLAGS_HEAPTYPE flag and are garbage collected.
+ * Our DTypeMeta instances (`np.dtype` and its subclasses) *may* be HeapTypes
+ * if the Py_TPFLAGS_HEAPTYPE flag is set (they are created from Python).
+ * They are not for legacy DTypes or np.dtype itself.
+ *
+ * @param self
+ * @return nonzero if the object is garbage collected
+ */
+static NPY_INLINE int
+dtypemeta_is_gc(PyObject *dtype_class)
+{
+    return PyType_Type.tp_is_gc(dtype_class);
+}
+
+
+static int
+dtypemeta_traverse(PyArray_DTypeMeta *type, visitproc visit, void *arg)
+{
+    /*
+     * We have to traverse the base class (if it is a HeapType).
+     * PyType_Type will handle this logic for us.
+     * This function is currently not used, but will probably be necessary
+     * in the future when we implement HeapTypes (python/dynamically
+     * defined types). It should be revised at that time.
+     */
+    assert(0);
```

Comment:
Thanks for that and the other comments.  This one was intentional, because I thought if we actually start using it, we should check that it actually works correctly :).